### PR TITLE
Implement `client:only` handling

### PR DESCRIPTION
**Pull request by natemoo-re**
## Changes
To be paired with https://github.com/snowpackjs/astro-compiler-next/pull/128

- [x] Adds support back for `client:only` directive
- [x] Handle SSR gracefully (AKA do not SSR the relevant module)

## Testing

Tested in compiler

## Docs

N/A, compat with Astro Classic™

### CODE REVIEW

File: packages/astro/src/runtime/server/index.ts
Comment (drwpow) [inline packages/astro/src/runtime/server/index.ts:186]: 👍🏻

File: packages/astro/test/astro-client-only.test.js
Comment (drwpow) [inline packages/astro/test/astro-client-only.test.js:40]: Woohoo!

File: packages/astro/test/astro-dynamic.test.js
Comment (drwpow) [inline packages/astro/test/astro-dynamic.test.js:45]: Some other tests had a note about this being hard to locate with the move to Vite, along with a TODO. Maybe add one here?

File: packages/astro/src/runtime/server/index.ts
Comment (drwpow) [inline packages/astro/src/runtime/server/index.ts:129]: Does this still work if there’s only one renderer (above `if` check)? Seems like we need to just not set a renderer for `client:only` right?

Comment (natemoo-re) [reply packages/astro/src/runtime/server/index.ts:129]: If there's only one renderer, we don't need to loop through them to find a match in either case. The first renderer will always apply, even for `client:only`.

**Comment by natemoo-re** [comment]
Please see https://github.com/snowpackjs/astro-compiler-next/pull/140 for a change that needs to happen in the compiler.

File: packages/astro/src/runtime/server/metadata.ts
Comment (jonathantneal) [inline packages/astro/src/runtime/server/metadata.ts:21]: 👍

File: packages/astro/test/fixtures/custom-elements/my-component-lib/server.js
Comment (jonathantneal) [inline packages/astro/test/fixtures/custom-elements/my-component-lib/server.js:23]: Will this ever be false if `check` has already guarunteed it?

**Review by jonathantneal — COMMENTED** [review]
Only a minor nit on the check in `renderToStaticMarkup`. If things slip thru `check`, I’d like to understand how that happens and perhaps get that fixed.

File: packages/astro/test/fixtures/custom-elements/my-component-lib/server.js
Comment (natemoo-re) [reply packages/astro/test/fixtures/custom-elements/my-component-lib/server.js:23]: Good question, this was a change that was breaking the `custom-elements` tests! Really unclear why this was needed, but `check` seemed to be passing even for client-only components due to the way polyfills were being loaded.

I'd like some more insight here, too. Did I break something or is this expected? Would love if @matthewp could take a look.

Comment (matthewp) [reply packages/astro/test/fixtures/custom-elements/my-component-lib/server.js:23]: Seems like your change shouldn't affect this. Would need to debug to see what's going on

**Comment by jonathantneal** [comment]
Please correct me if I’m wrong —

This PR is held up by an unrelated bug uncovered by @natemoo-re. The bug seems to be that when a component does not pass `check` it might still be processed by `renderComponent`, which would fail. In this PR, @natemoo-re has added a work-around to the web components test to avoid the failure and our test suite keeps passing.

Should we allow the work-around so that this can be merged?
Or, should we wait for the uncovered bug to be resolved before this is merged?

Personally, I would be in favor of merging this before the fix if 1. the investigation and/or fix by @matthewp could take long enough that this PR gets stale, and 2. @natemoo-re would be okay adding a big TODO comment above the check in `renderComponent`. Thoughts?

**Comment by matthewp** [comment]
I don't think we should merge this. It seems that it breaks existing functionality, otherwise the condition wouldn't be needed. It could very well break real code, we don't know for sure yet. I still need to investigate what is going on.

**Comment by natemoo-re** [comment]
I agree with @matthewp's assessment. I suspect this could have been caused by some of my changes to the `runtime/server/index.ts` file, but I haven't had the time to dig back into it. Appreciate any help doing so ☺️

**Comment by matthewp** [comment]
I'm going to hand this one back to you. I wasn't able to get very far on this due to not being able to run these tests locally due to [this issue](https://github.com/snowpackjs/astro/blob/b49f9a525e5eb9f497259a50c73af9e370f618be/packages/astro/src/vite-plugin-astro/index.ts#L85-L87).

That was introduced in this commit: https://github.com/snowpackjs/astro-compiler-next/commit/6041ee577b4e22e65882c1027164a2bbfb3dea6e

Not sure the exact reason, but locally (on main) you can recreate by doing:

1. Checkout `@astrojs/compiler@0.2.24`
2. `make astro-wasm`
3. cd into your `astro` repo
4. `cp ~/go/src/github.com/snowpackjs/astro-compiler-next/lib/compiler/astro.wasm node_modules/@astrojs/compiler/`
5. `node node_modules/.bin/_mocha packages/astro/test/custom-elements.test.js`

Which should pass. Then cherry-pick eat of the commits in order https://github.com/snowpackjs/astro-compiler-next/commits/main repeating the same process. They all should pass, until you get to 6041ee577b4e22e65882c1027164a2bbfb3dea6e, in which it fails.

It's very possible that the compiler issue is the reason for this workaround as well.

**Comment by matthewp** [comment]
Fix for the underlying compiler thing is here: https://github.com/snowpackjs/astro-compiler-next/pull/146

**Review by matthewp — APPROVED** [review]
lgtm

**Pull request merged**
